transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/persistence/SQLiteEventStore.java

```diff
@@ -531,6 +538,161 @@ private static Encoding toEncoding(@Nullable String value) {
     } while (true);
   }
 
+  /**
+   * Record log event dropped.
+   *
+   * @param eventsDroppedCount number of events dropped
+   * @param reason reason why events are dropped
+   * @param logSource log source of the dropped events
+   */
+  @Override
+  public void recordLogEventDropped(
+      long eventsDroppedCount, LogEventDropped.Reason reason, String logSource) {
+    if (!isLogEventDroppedExist(reason, logSource)) {
+      ContentValues metrics = new ContentValues();
+      metrics.put("log_source", logSource);
+      metrics.put("reason", reason.getNumber());
+      metrics.put("events_dropped_count", eventsDroppedCount);
+      inTransaction(
+          db -> {
+            db.insert("log_event_dropped", null, metrics);
+            return null;
+          });
+    } else {
+      String query =
+          "UPDATE log_event_dropped SET events_dropped_count = events_dropped_count + "
+              + eventsDroppedCount
+              + " WHERE log_source = ? AND reason = ?";
+      inTransaction(
+          db -> {
+            db.execSQL(query, new String[] {logSource, Integer.toString(reason.getNumber())});
+            return null;
+          });
+    }
+  }
+
+  private LogEventDropped.Reason convertToReason(int number) {
+    if (number == LogEventDropped.Reason.REASON_UNKNOWN.getNumber()) {
+      return LogEventDropped.Reason.REASON_UNKNOWN;
+    } else if (number == LogEventDropped.Reason.MESSAGE_TOO_OLD.getNumber()) {
+      return LogEventDropped.Reason.MESSAGE_TOO_OLD;
+    } else if (number == LogEventDropped.Reason.CACHE_FULL.getNumber()) {
+      return LogEventDropped.Reason.CACHE_FULL;
+    } else if (number == LogEventDropped.Reason.PAYLOAD_TOO_BIG.getNumber()) {
+      return LogEventDropped.Reason.PAYLOAD_TOO_BIG;
+    } else if (number == LogEventDropped.Reason.MAX_RETRIES_REACHED.getNumber()) {
+      return LogEventDropped.Reason.MAX_RETRIES_REACHED;
+    } else if (number == LogEventDropped.Reason.INVALID_PAYLOD.getNumber()) {
+      return LogEventDropped.Reason.INVALID_PAYLOD;
+    } else if (number == LogEventDropped.Reason.SERVER_ERROR.getNumber()) {
+      return LogEventDropped.Reason.SERVER_ERROR;
+    } else {
```

Comment:
This whole block is a good candidate for a `switch()`, any reason you're not using it here?

Reply:
we cannot use `switch` here, since `LogEventDropped.Reason.INVALID_PAYLOD.getNumber()` is not an constant expression. If we go back to use `switch`, we will have to do `case 0:` `case 1:`, but i feel using `LogEventDropped.Reason.INVALID_PAYLOD.getNumber()` with `if else` is more clear.

Reply:
sgtm